### PR TITLE
[FLINK-29042][Connectors/ElasticSearch] Support lookup join for es connector

### PR DESCRIPTION
**Pull request by liyubin117**
Now es connector could only be used as a sink, but in many business scenarios, we treat es as a index database, thus we should support to make it lookupable in flink.

### CODE REVIEW

**Comment by liyubin117** [comment]
@MartijnVisser Hi. Could you please give a review? Thanks!

**Comment by MartijnVisser** [comment]
@vahmed-hamdy Since you offered to volunteer to review an Elasticsearch PR, perhaps you could also have a look at this open PR?

**Comment by liyubin117** [comment]
@MartijnVisser @vahmed-hamdy Hi, The feature has taken effect in our production environment for one more years, now we want to updrade our es connector, if the feature would be merged, it would be very nice to upgrade directly!

**Comment by MartijnVisser** [comment]
@reta Do you have any considerations before merging this?

**Comment by reta** [comment]
> @reta Do you have any considerations before merging this?

@MartijnVisser LGTM, we would probably could bring that to OpenSearch connector as well, thank you!

File: flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                        RowData row = parseSearchResult(s);
+                        rows.add(row);
+                    }
+                    rows.trimToSize();
+                    return rows;
+                }
+            } catch (IOException e) {
+                LOG.error(String.format("Elasticsearch search error, retry times = %d", retry), e);
+                if (retry >= maxRetryTimes) {
+                    throw new RuntimeException("Execution of Elasticsearch search failed.", e);
```
Comment (snuyanzin) [inline flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java:158]:
```suggestion
                    throw new FlinkRuntimeException("Execution of Elasticsearch search failed.", e);
```

File: flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                }
+                try {
+                    Thread.sleep(1000L * retry);
+                } catch (InterruptedException e1) {
+                    LOG.warn(
+                            "Interrupted while waiting to retry failed elasticsearch search, aborting");
+                    throw new RuntimeException(e1);
```
Comment (snuyanzin) [inline flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java:165]:
```suggestion
                    throw new FlinkRuntimeException(e1);
```

File: flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** A lookup function implementing {@link LookupTableSource} in elasticsearch connector. */
```
Comment (snuyanzin) [inline flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/ElasticsearchRowDataLookupFunction.java:52]:
```suggestion
/** A lookup function implementing {@link LookupTableSource} in Elasticsearch connector. */
```

File: flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSource.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Override
+    public DynamicTableSource copy() {
+        return new Elasticsearch6DynamicSource(
+                format, config, physicalRowDataType, lookupMaxRetryTimes, lookupCache);
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "Elasticsearch-6";
```
Comment (snuyanzin) [inline flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSource.java:123]:
```suggestion
        return "Elasticsearch6";
```
There is DynamicSink with `Elasticsearch6`, shouldn't we follow same naming conventions?

File: flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+
+    @Override
+    public DynamicTableSource copy() {
+        return new Elasticsearch7DynamicSource(
+                format, config, physicalRowDataType, lookupMaxRetryTimes, lookupCache);
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "Elasticsearch-6";
```
Comment (snuyanzin) [inline flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java:123]:
```suggestion
        return "Elasticsearch6";
```

File: flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                            config.getPathPrefix().orElse(null),
+                            config.getUsername().get(),
+                            config.getPassword().get());
+        } else {
+            restClientFactory =
+                    new Elasticsearch7DynamicSink.DefaultRestClientFactory(
+                            config.getPathPrefix().orElse(null));
+        }
+
+        Elasticsearch7ApiCallBridge elasticsearch6ApiCallBridge =
```
Comment (snuyanzin) [inline flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java:86]:
```suggestion
        Elasticsearch7ApiCallBridge elasticsearch7ApiCallBridge =
```
shouldn't it be 7?

File: flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        ElasticsearchRowDataLookupFunction<RestHighLevelClient> lookupFunction =
+                new ElasticsearchRowDataLookupFunction<>(
+                        this.format.createRuntimeDecoder(context, physicalRowDataType),
+                        lookupMaxRetryTimes,
+                        config.getIndex(),
+                        config.getDocumentType(),
+                        DataType.getFieldNames(physicalRowDataType).toArray(new String[0]),
+                        DataType.getFieldDataTypes(physicalRowDataType).toArray(new DataType[0]),
+                        keyNames,
+                        elasticsearch6ApiCallBridge);
```
Comment (snuyanzin) [inline flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java:107]:
```suggestion
                        elasticsearch7ApiCallBridge);
```

**Comment by snuyanzin** [comment]
thanks for the contribution
also looks good from my side
i left some minor comments

+1 for having similar things for OpenSearch

File: flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSource.java (hunk 4 above)
Comment (liyubin117) [reply flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSource.java:123]: wise consideration 👍

File: flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java (hunk 5 above)
Comment (liyubin117) [reply flink-connector-elasticsearch7/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch7DynamicSource.java:123]: yeah, it should be 7

**Comment by MartijnVisser** [comment]
@snuyanzin Will you take care of the rest of the review and merge it when you think it's OK?

**Comment by snuyanzin** [comment]
yes, sure, will try to do it later today/tomorrow

**Comment by liyubin117** [comment]
CI failed caused by outdated dev branch, I would rebase on the latest main branch.

**Review by snuyanzin — APPROVED** [review]
Thanks for addressing comments

**Comment by liyubin117** [comment]
@snuyanzin  @MartijnVisser Thanks a lot for your patient work.

**Pull request merged**